hypothesis-python/src/hypothesis/extra/numpy.py

```diff
@@ -133,42 +133,18 @@ def __init__(self, element_strategy, shape, dtype, fill, unique):
         self.element_strategy = element_strategy
         self.unique = unique
 
-        # Used by self.insert_element to check that the value can be stored
-        # in the array without e.g. overflowing.  See issues #1385 and #1591.
-        if dtype.kind in (u"i", u"u"):
-            self.check_cast = lambda x: np.can_cast(x, self.dtype, "safe")
-        elif dtype.kind == u"f" and dtype.itemsize == 2:
-            max_f2 = (2.0 - 2 ** -10) * 2 ** 15
-            self.check_cast = lambda x: (not np.isfinite(x)) or (-max_f2 <= x <= max_f2)
-        elif dtype.kind == u"f" and dtype.itemsize == 4:
-            max_f4 = (2.0 - 2 ** -23) * 2 ** 127
-            self.check_cast = lambda x: (not np.isfinite(x)) or (-max_f4 <= x <= max_f4)
-        elif dtype.kind == u"c" and dtype.itemsize == 8:
-            max_f4 = (2.0 - 2 ** -23) * 2 ** 127
-            self.check_cast = lambda x: (not np.isfinite(x)) or (
-                -max_f4 <= x.real <= max_f4 and -max_f4 <= x.imag <= max_f4
-            )
-        elif dtype.kind == u"U":
-            length = dtype.itemsize // 4
-            self.check_cast = lambda x: len(x) <= length and u"\0" not in x[length:]
-        elif dtype.kind in (u"S", u"a"):
-            self.check_cast = (
-                lambda x: len(x) <= dtype.itemsize and b"\0" not in x[dtype.itemsize :]
-            )
-        else:
-            self.check_cast = lambda x: True
-
     def set_element(self, data, result, idx, strategy=None):
         strategy = strategy or self.element_strategy
         val = data.draw(strategy)
         result[idx] = val
-        if self._report_overflow and not self.check_cast(val):
+        if self._report_overflow and val != result[idx] and val == val:
```

Comment:
`val != result[idx] and val == val` - what's the case for val when it's not equal and not not equal to itself? Except NaN. If NaN is the only one, may be check against it explicitly?

Reply:
`float("nan")` is not equal to itself!

So we only care that val does not equal the element we just inserted into the array if val is equal to itself, in order to skip the warning for nan.

Reply:
Yes, I'm originally wrote about NaN, but it isn't a case since NaN not equat to NaN and not not equal to NaN.
```
>>> float('nan') != float('nan') and float('nan') == float('nan')
False
```
So it'll never pass this condition. What the else remains?

Reply:
Oh, sorry - `val != result[idx]` can be true because of integer wraparound, converting 64 to 32 (or 16) bit float, truncating a string at the first null byte...

Basically because Numpy representations are more machine-level, so it's easy to generate elements that can't quite be represented in the array.